OpenSim/Common/ComponentPath.h

```diff
@@ -0,0 +1,90 @@
+ * Copyright (c) 2005-2016 Stanford University and the Authors                *
+ * Author(s): Carmichael Ong                                                  *
+ *                                                                            *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may    *
+ * not use this file except in compliance with the License. You may obtain a  *
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+#include "Path.h"
+
+namespace OpenSim {
+
+//==============================================================================
+//                            OPENSIM COMPONENT PATH
+//==============================================================================
+/**
+* A class for handling Paths for Components, deriving from Path. A ComponentPath
+* uses a forward-slash ('/') as a separator always. It also specifies invalid
+* characters in a Component name:
+* - back-slash ('\\')
+* - forward-slash ('/')
+* - asterisk ('*')
+* - plus-sign ('+')
+*
+* @author Carmichael Ong
+*/
+
+class OSIMCOMMON_API ComponentPath : public Path {
+    OpenSim_DECLARE_CONCRETE_OBJECT(ComponentPath, Path);
+
+public:
+    /// Constructors
+    ComponentPath() = default;
+    ComponentPath(const std::string path);
+    ComponentPath(std::vector<std::string> pathVec, bool isAbsolute);
+
+    /// Operators
+    bool operator==(const ComponentPath& other) const
+    {
+        return this->getString() == other.getString();
+    }
+
+    bool operator!=(const ComponentPath& other) const
+    {
+        return this->getString() != other.getString();
+    }
+
+    /// Get an absolute path by resolving it relative to a given otherPath.
+    /// If the current Path is already absolute, return the same Path.
+    ComponentPath getAbsolutePath(ComponentPath* otherPath);
+
+    /// Find the relative Path between this Path and another Path (otherPath)
+    /// (i.e. the Path to go FROM otherPath TO this Path). Both Paths must be 
+    /// absolute.
+    ComponentPath getRelativePath(ComponentPath* otherPath);
+
+    /// Return the the sub-path that contains all sub components except for
+    /// the last one.
+    ComponentPath getParentPath();
+
+    /// Return a ComponentPath object of a subcomponent by index. This will
+    /// always be a relative path.
+    ComponentPath getSubcomponent(size_t index);
```

Comment:
> Unfortunately "Path" is highly overloaded. That is why the line "path" for muscles, ligaments, etc.. is a `GeometryPath` Component and see the confusion with using "PathComponent".

The word "path" is clear; what is not immediately obvious (to me, at least) is what you mean by "element" or "component" since AFAIK these terms deviate from the terminology typically used to discuss paths. I suggest using established terminology if possible.

I would either use the standard *file system* terminology throughout (e.g., `getSubdirectoryNameAtLevel()`) with the understanding that the Component tree is directly analogous to a file system, or use standard *tree data structure* terminology throughout (e.g., `getSubtreeNodeNameAtLevel()`).